include/boost/gil/image.hpp

```diff
@@ -122,6 +135,48 @@ class image {
         return *this;
     }
 
+private:
+    using equal_allocators = std::true_type;
+    using no_propagate_allocators = std::false_type;
+
+    template <class Allocator>
+    // TODO: Use std::allocator_traits<Allocator>::is_always_equal if available
+    using move_policy = typename std::is_empty<Allocator>::type;
+
+    void move_assign(image& img, equal_allocators) {
+        destruct_pixels(_view);
+        deallocate();
+
+        // TODO Use std::exchange
+        _view = img._view;
+        _memory = img._memory;
+        _align_in_bytes = img._align_in_bytes;
+        _allocated_bytes = img._allocated_bytes;
+        
+        img._view = view_t();
+        img._memory = nullptr;
+        img._align_in_bytes = 0;
+        img._allocated_bytes = 0;
+    }
```

Comment:
I have to admit I still have to catch up with the hairy topic of the allocators and related container behaviours in post the C++03 era, but what I am missing in the whole move-assignment is this :

- use of `propagate_on_container_move_assignment` trait
- `// fallback to copy nothing` see below

This below is complete illustration what I think we need, at least according to my current mis-or-understanding of this matter:

```cpp
image& operator=(image&& rhs)
{
    auto const exchange_memory = [](image& lhs, image& rhs)
    {
        lhs._memory = std::exchange(rhs._memory, nullptr);
        lhs._align_in_bytes = std::exchange(rhs._align_in_bytes, 0);
        lhs._allocated_bytes = std::exchange(rhs._allocated_bytes, 0);
        lhs._view = std::exchange(rhs._view, image::view_t{});
    };

    constexpr bool pocma = std::allocator_traits<Alloc>::propagate_on_container_move_assignment::value;
    if (pocma)
    {
        // non-sticky allocator, can adopt the memory, fast
        destruct_pixels(this->_view);
        this->deallocate();
        this->_alloc = rhs._alloc;
        exchange_memory(*this, rhs);
    }
    else if (_alloc == rhs._alloc)
    {
        // allocator stuck to the rhs, but it's equivalent of ours, we can still adopt the memory
        destruct_pixels(_view);
        this->deallocate();
        exchange_memory(*this, rhs);
    }
    else
    {
        // cannot propagate the allocator and cannot adopt the memory
        if (rhs._memory)
        {
            allocate_and_copy(rhs.dimensions(), rhs._view);
            destruct_pixels(rhs._view);
            rhs.deallocate();
            rhs._view = image::view_t{};
        }
        else
        {
            destruct_pixels(this->_view);
            this->deallocate();
            this->_view = view_t{};
        }
    }
    return *this;
}
```

What do you think @sdebionne & @mocabe ?

Reply:
I am not really qualified I am afraid! So I basically checked how standard libraries (the one from MS) implement move semantic for `std::vector`:

```
struct equal_allocators {};
using propagate_allocators = std::true_type;
using no_propagate_allocators = std::false_type;

template <class Alloc>
using pocma  = conditional_t<
  std:allocator_traits<Alloc>::is_always_equal::value,
  equal_allocators,
  typename std::allocator_traits<Alloc>::propagate_on_container_move_assignment::type
>;
```

We don't have `is_always_equal` in C++11 but could replace it with `is_empty`. If we use `propagate_on_container_move_assignment` only, then your implementation looks good to me.

Your if statement is much better than my template specialization.

Reply:
It is a good idea to check what implementors do.
I also see MSVC++ STL does this:

```cpp
template <class _Alloc>
void _Pocma(_Alloc& _Left, _Alloc& _Right) noexcept {
    if constexpr (allocator_traits<_Alloc>::propagate_on_container_move_assignment::value) {
        _Left = _STD move(_Right);
    } else {
        (void) _Left;
        (void) _Right;
    }
}
```

which is equivalent to this if-else distinguishing between these two in my version above:

```cpp
if (pocma)
{
    this->_alloc = rhs._alloc; // we could std::move(rhs._alloc) here
    exchange_memory(*this, rhs);
}
else if (_alloc == rhs._alloc)
{
    // no-op for rhs._alloc
    this->deallocate();
    exchange_memory(*this, rhs);
}
```

> Your if statement is much better than my template specialization.

I just find the locality of all the aspects make it readable.

I think the metafunctions in the MSVC++ STL are good idea as soon as we need to re-use some code in number of places, as STL does for various containers.

There is one place they will be useful, for handling of `noexcept` which in STL is conditional this way:

```cpp
noexcept(noexcept(_Move_assign(_Right, _Choose_pocma<_Alty>{})))
```

Reply:
> which is equivalent to this if-else distinguishing between these two in my version above  

I looks like STL's implementation actually does not call `operator=` on allocator object when it cannot be propagated on move assignment, which allows allocators to define `operator=` deleted.

Reply:
@mocabe Yes, indeed, that's also what we do in the snippet above: in `// allocator stuck to the rhs,` and the following `else`-cases. IOW, I consider STL's and the unrolled `if-else` versions equivalent.

Reply:
@mloskot I mean, `if-else` version requires definition of `operator=` of allocator type at compile time. STL uses `if constexpr` so it will discard instantiation of `_Alloc::operator=` when the statement is not active.

Reply:
@mocabe Right. I've got too much into the C++11 mode. This is worth to consider now as future-proof indeed. @sdebionne What you think? If you don't have time, I will try to go back to your variant, tweaked, later this weak.